proposals/0132-sequence-end-ops.md

```diff
@@ -16,724 +17,227 @@ these names so they follow consistent, predictable patterns.
 
-|                                                  | Type                         | Get                       | Remove                              | Replace                                                     |
-| ------------------------------------------------ | ---------------------------- | ------------------------- | ----------------------------------- | ----------------------------------------------------------- |
-| **Based&nbsp;on&nbsp;Index,&nbsp;Arbitrary**     |
-| (i: Index) ..< (j: Index)                        | Range\<Index>                | C[i ..< j]                | C.removeSubrange(i&nbsp;..<&nbsp;j) | C.replaceSubrange(i&nbsp;..<&nbsp;j,&nbsp;with:&nbsp;[x,y]) |
-| &nbsp;&nbsp;...Countable                         | CountableRange\<Index>       | C[i ..< j]                | C.removeSubrange(i&nbsp;..<&nbsp;j) | C.replaceSubrange(i&nbsp;..<&nbsp;j,&nbsp;with:&nbsp;[x,y]) |
-| (i: Index) ... (j: Index)                        | ClosedRange\<Index>          | C[i ... j]                | C.removeSubrange(i ... j)           | C.replaceSubrange(i ... j, with: [x,y])                     |
-| &nbsp;&nbsp;...Countable                         | CountableClosedRange\<Index> | C[i ... j]                | C.removeSubrange(i ... j)           | C.replaceSubrange(i ... j, with: [x,y])                     |
-| **Based&nbsp;on&nbsp;Index,&nbsp;From&nbsp;End** |
-| startIndex ..< (i: Index)                        | upTo: Index                  | C.prefix(upTo:&nbsp;i)    | -                                   | -                                                           |
-| (i: Index) ..< endIndex                          | from: Index                  | C.suffix(from:&nbsp;i)    | -                                   | -                                                           |
-| startIndex ... (i: Index)                        | through: Index               | C.prefix(through:&nbsp;i) | -                                   | -                                                           |
-
-### Why does it matter?
-
-Many of these APIs are only occasionally necessary, so it's important 
-that they be easy to find when needed and easy to understand when 
-read. If you know that `prefix(10)` will get the first ten elements but 
-don't know what its inverse is, you will probably not guess that it's 
-`dropFirst(10)`. The confusing, conflicting names in these APIs are a 
-barrier to users adopting them where appropriate.
+  [onesided]: (0172-one-sided-ranges.md)
+  
+| Operation        | Operand                           | Current name                  |
+| ---------------- | --------------------------------- | ----------------------------- |
+| Get value        | First element                     | `first`                       |
+| "                | Last element                      | `last`                        |
+| "                | First *n* elements                | `prefix(3)`                   |
```

Comment:
I don’t think of this as a “get value” operation at all.  This is slicing, and should be spelled like other slicing operations.  I’ve posted other syntax suggestions in the past; here are a few more, just to get the juices flowing.

```swift
x[...^3] // first 3 elements
x[3$...] // last 3 elements
```

Reply:
...surely you mean the first example to be a spelling for retrieving _last_ three elements, and the second to be the spelling for retrieving the _first_ three?

(I think this is where words like `first` and `prefix` really improve readability. As a pragmatic matter I'd value that over the purity of all slicing operations having the same notation.)

Reply:
I’m on phone only so it’d be difficult to check, but aren’t `prefix(_:)` and friends on `Sequence`? Even if we want to convert these into punctuation, we would be extending subscripts into places they are not currently used.

Reply:
@xwu, no I meant it the way I wrote it, with `^`*N*/*N*`^` indicating “*N* places from the start” and `$`*N*/*N*`$` indicating “*N* places from the end.”  If you read it that way, the whole thing fits together with `x[i...]`/`x[...j]`. The fact that it confused you might indicate that this isn’t the right notational system, or it might just be something you need to learn to read once.  And, uh, the first one should have been `x[..<^3]` or `X[...^2]` I suppose :(.

Reply:
@dabrahams Unless there are six (or is it five?--zero-based counting is hard) elements, why would `x[..<^3]`, meaning "three places from the end," also mean "first three elements"?

Reply:
The caret means “from the beginning.”  To say ‘from the end” you use the dollar sign.

Reply:
I see. So `x[..<3]` _always_ gives you three, and only if the index type is an integer, but `x[..<^3]` gives you at most three, and even if the index type is opaque?